src/ipsecondarydiscovery.cc

```diff
@@ -0,0 +1,94 @@
+#include "ipsecondarydiscovery.h"
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <chrono>
+
+#include "logging.h"
+
+std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::discover() {
+  sendRequest();
+  return waitDevices();
+}
+
+std::vector<Uptane::SecondaryConfig> IpSecondaryDiscovery::waitDevices() {
+  std::vector<Uptane::SecondaryConfig> secondaries;
+  int socket_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (socket_fd == -1) {
+    perror("socket creation");
+  }
+
+  int reuse = 1;  // Neded for tests, because client and server runs on the same machine and bind the same port
+  setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof reuse);
+
+  struct sockaddr_in recv_addr {};
+  recv_addr.sin_family = AF_INET;
+  recv_addr.sin_port = htons(config_.ipdiscovery_port);
+  recv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+  if (bind(socket_fd, (struct sockaddr *)&recv_addr, sizeof recv_addr) < 0) {
+    perror("bind");
+  }
+
+  auto start_time = std::chrono::system_clock::now();
+  struct timeval tv;
+  tv.tv_sec = config_.ipdiscovery_wait_seconds;
+  tv.tv_usec = 0;
+  setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+
+  char rbuf[2000] = {};
+  int recieved = 0;
+  while ((recieved = recv(socket_fd, rbuf, sizeof(rbuf) - 1, 0)) != -1) {
+    std::string data(std::move(rbuf), recieved);
+    try {
+      int type = 0;
+      Uptane::SecondaryConfig conf;
+      asn1::Deserializer des(data);
+      des >> asn1::seq >> asn1::implicit<kAsn1Integer>(type) >> asn1::implicit<kAsn1Utf8String>(conf.ecu_serial) >>
+          asn1::implicit<kAsn1Utf8String>(conf.ecu_hardware_id) >> asn1::restseq;
+      if (type == AKT_DISCOVERY_RESP) {
+        conf.secondary_type = Uptane::SecondaryType::kUptane;
+        secondaries.push_back(conf);
+      }
+    } catch (const deserialization_error ex) {
+      LOG_ERROR << ex.what();
+    }
+    auto now = std::chrono::system_clock::now();
+    int left_seconds =
+        config_.ipdiscovery_wait_seconds - std::chrono::duration_cast<std::chrono::seconds>(now - start_time).count();
+    if (left_seconds <= 0) {
+      break;
+    }
+    tv.tv_sec = left_seconds;
+    setsockopt(socket_fd, SOL_SOCKET, SO_RCVTIMEO, (const char *)&tv, sizeof(tv));
+  }
+  return secondaries;
+}
+
+void IpSecondaryDiscovery::sendRequest() {
+  int socket_fd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (socket_fd == -1) {
+    LOG_ERROR << "Error of creating socket: " << std::strerror(errno);
```

Comment:
The function should return or throw in case of errors (also in subsequent `setsockopt`).

To make that easier, I was about to move `SocketHandle` to `utils.h`:
https://github.com/advancedtelematic/aktualizr/blob/05eeb37d2bfaa0e31cc26a4a6c4d610efd7f8ba9/src/aktualizr_secondary/aktualizr_secondary.h#L17
But no need to use it right away.